From 85c28938cf51fd88831927f037753624fb974913 Mon Sep 17 00:00:00 2001
From: Tan Siewert <tan.siewert@9elements.com>
Date: Thu, 23 Jul 2026 17:45:50 +0200
Subject: [PATCH] Add HPE GXP

Add the minimum requirement for HPE GXP Platform Init and full support
for CMOS reset.

Tested: Code compiles and CMOS reset works

Upstream-Status: Pending
Change-Id: I59dfb0491102468af8cbad22c80f544f366d0c57
Signed-off-by: Tan Siewert <tan.siewert@9elements.com>
---
 meson.build                 |   5 ++
 meson.options               |   1 +
 platform.hpp.in             |   1 +
 platforms/hpe/gxp.cpp       | 116 ++++++++++++++++++++++++++++++++++++
 platforms/hpe/platforms.hpp |  15 +++++
 5 files changed, 138 insertions(+)
 create mode 100644 platforms/hpe/gxp.cpp
 create mode 100644 platforms/hpe/platforms.hpp

diff --git a/meson.build b/meson.build
index 9a4070e..00e0cf1 100644
--- a/meson.build
+++ b/meson.build
@@ -42,6 +42,11 @@ platforms_map = {
         'src': files('platforms/intel/jcrp.cpp'),
         'fn': 'intel::init_jcrp',
     },
+    'hpe-gxp': {
+        'src': files('platforms/hpe/gxp.cpp'),
+        'fn': 'hpe::init_gxp_base',
+        'cmos_reset_fn': 'hpe::gxp_cmos_reset',
+    },
     'nvidia-gb200': {
         'src': files('platforms/nvidia/gb200.cpp'),
         'fn': 'nvidia::init_gb200_base',
diff --git a/meson.options b/meson.options
index 8f4ea77..74fbdf5 100644
--- a/meson.options
+++ b/meson.options
@@ -4,6 +4,7 @@ option(
     choices: [
         'intel-acrp',
         'intel-jcrp',
+        'hpe-gxp',
         'nvidia-gb200',
         'nvidia-gb200-with-p2020',
         'nvidia-nvl32',
diff --git a/platform.hpp.in b/platform.hpp.in
index 91350df..8a77910 100644
--- a/platform.hpp.in
+++ b/platform.hpp.in
@@ -4,6 +4,7 @@
 #pragma once
 
 #include "intel/platforms.hpp"
+#include "hpe/platforms.hpp"
 #include "nvidia/platforms.hpp"
 
 #include <sdbusplus/async.hpp>
diff --git a/platforms/hpe/gxp.cpp b/platforms/hpe/gxp.cpp
new file mode 100644
index 0000000..65faddc
--- /dev/null
+++ b/platforms/hpe/gxp.cpp
@@ -0,0 +1,116 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright OpenBMC Authors
+
+#include "platforms.hpp"
+#include "utilities.hpp"
+
+#include <systemd/sd-daemon.h>
+
+#include <sdbusplus/asio/connection.hpp>
+
+#include <chrono>
+#include <filesystem>
+#include <format>
+#include <fstream>
+#include <iostream>
+#include <string_view>
+
+namespace hpe
+{
+
+constexpr const auto nvramPath = "/dev/mtd/host-reserved";
+constexpr const auto chifEvsPath = "/var/lib/chif/evs.dat";
+constexpr const auto chifSvc = "xyz.openbmc_project.GxpChif.service";
+
+constexpr const auto systemdService = "org.freedesktop.systemd1";
+constexpr const auto systemdObjPath = "/org/freedesktop/systemd1";
+constexpr const auto systemdManagerInterface =
+    "org.freedesktop.systemd1.Manager";
+
+bool restartUnit(sdbusplus::bus_t& bus, const std::string& unit)
+{
+    lg2::info("Restarting {UNIT}", "UNIT", unit);
+
+    auto method = bus.new_method_call(systemdService, systemdObjPath,
+                                      systemdManagerInterface, "RestartUnit");
+
+    method.append(unit);
+    method.append("replace");
+
+    try
+    {
+        bus.call_noreply(method);
+    }
+    catch (const sdbusplus::exception_t& e)
+    {
+        lg2::error("Failed to restart unit {UNIT}, exception: {ERR}", "UNIT",
+                   unit, "ERR", e);
+        return false;
+    }
+
+    return true;
+}
+
+int logged_system(std::string_view cmd)
+{
+    lg2::info("calling {CMD}", "CMD", cmd);
+    int rc = std::system(cmd.data());
+    return rc;
+}
+
+sdbusplus::async::task<bool> gxp_cmos_reset(
+    [[maybe_unused]] sdbusplus::async::context& ctx)
+{
+    auto bus = sdbusplus::bus::new_default_system();
+
+    lg2::info("CMOS reset triggered");
+    wait_for_path_to_exist(nvramPath, std::chrono::seconds(10));
+
+    int rc = logged_system(std::format("flash_eraseall {}", nvramPath));
+
+    if (rc < 0)
+    {
+        lg2::error("NVRAM erase failed with {ERR}", "ERR", rc);
+        co_return false;
+    }
+
+    try
+    {
+        if (std::filesystem::remove(chifEvsPath))
+        {
+            lg2::info("EV storage cleared");
+        }
+        else
+        {
+            lg2::warning("EV storage not found. Maybe removed?");
+        }
+    }
+    catch (const std::filesystem::filesystem_error& err)
+    {
+        lg2::error("Error clearing EV storage: {ERR}", "ERR", err.what());
+        co_return false;
+    }
+
+    if (!restartUnit(bus, chifSvc))
+    {
+        co_return false;
+    }
+
+    co_return true;
+}
+
+int init_gxp_base()
+{
+    lg2::info("GXP platform init");
+
+    int notify_rc = sd_notify(0, "READY=1\nSTATUS=GXP init complete");
+
+    if (notify_rc < 0)
+    {
+        lg2::error("sd_notify READY=1 failed: {RC}", "RC", notify_rc);
+        return EXIT_FAILURE;
+    }
+
+    return EXIT_SUCCESS;
+}
+} // namespace hpe
diff --git a/platforms/hpe/platforms.hpp b/platforms/hpe/platforms.hpp
new file mode 100644
index 0000000..28fd716
--- /dev/null
+++ b/platforms/hpe/platforms.hpp
@@ -0,0 +1,15 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright OpenBMC Authors
+
+#pragma once
+
+#include <phosphor-logging/lg2.hpp>
+#include <sdbusplus/async.hpp>
+
+namespace hpe
+{
+
+int init_gxp_base();
+sdbusplus::async::task<bool> gxp_cmos_reset(sdbusplus::async::context&);
+
+} // namespace hpe
